app/assets/src/components/views/amr_heatmap/AMRHeatmapVis.jsx

```diff
@@ -0,0 +1,262 @@
+import React from "react";
+import PropTypes from "prop-types";
+import Heatmap from "~/components/visualizations/heatmap/Heatmap";
+import { getAMRCounts } from "~/api/amr";
+
+import cs from "./amr_heatmap_vis.scss";
+
+const METRIC_ALLELES = "alleles";
+const METRIC_GENES = "genes";
+
+export default class AMRHeatmapVis extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      viewLevel: "genes",
+      metric: "coverage",
+    };
+
+    this.heatmap = null;
+  }
+
+  componentDidMount() {
+    this.requestAMRCountsData(this.props.sampleIds);
+  }
+
+  componentDidUpdate() {
+    let rows = this.createHeatmapLabels();
+    let columns = this.state.samples;
+    let values = this.computeHeatmapValues(rows);
+    let options = {};
+    this.renderHeatmap(rows, columns, values, options);
+  }
+
+  async requestAMRCountsData(sampleIds) {
+    let rawResponse = await getAMRCounts(sampleIds);
+    let [geneData, samples] = this.extractData(rawResponse);
+    let alleleToGeneMap = this.mapAllelesToGenes(geneData);
+    this.setState((state, props) => {
+      return {
+        rawData: rawResponse,
+        geneData: geneData,
+        samples: samples,
+        alleleToGeneMap: alleleToGeneMap,
+      };
+    });
+  }
+
+  extractData(rawData) {
+    let genes = {};
+    let sampleLabels = [];
+    rawData.filter(rawSample => rawSample.error === "").forEach(sample => {
```

Comment:
The `genes` data structure seems overly complex. It's possible to implement `computeHeatmapValues` and `computeGeneLabels` with less overall code. You might not even need an intermediate data structure. 

I think the `lodash/fp` utility library will help you a lot. We have a small section on `lodash/fp` [here](https://github.com/chanzuckerberg/idseq-web/blob/master/DEV_GUIDELINES.md). The docs are [here](https://gist.github.com/jfmengels/6b973b69c491375117dc)

Here is some example code, for inspiration.

To get the unique gene labels:
```
// For each sample, map over each entry of sample.amr_counts, and get the "gene" field. 
// We use a lodash/fp shorthand of just specifying the field name "gene".
const getGeneLabelsForSample = sample => map("gene", sample.amr_counts)

// Map the above function over each sample.
const sampleGeneLabels = map(geneGeneLabelsForSample, rawData)
// Example: [["APH-Stph_AGly", "Dha1_Phe", "Tet-38_Tet"], ["APH-Stph_AGly", "Dha1_Phe", "Tet-38_Tet"]]

// Flatten the array and remove duplicates.
const uniqueGeneLabels = uniq(flatten(sampleGeneLabels))
// Example: ["APH-Stph_AGly", "Dha1_Phe", "Tet-38_Tet"]

// You can also do the above with one line. This is more compact but less readable.
// You should pick the option you think is most readable for the situation.
const uniqueGeneLabels = uniq(flatten(
  map(sample => map("gene", sample.amr_counts), rawData)
))
```

To determine, for a given gene label and sample name, whether there is a corresponding depth/coverage value (so that you can generate your heatmap values):

```
// Given sampleName, geneLabel (you'll need to map over all possible values)

// Get the matching sample object.
const sample = find(["sample_name", sampleName], rawData)

// Look for the matching amr count entry in the sample.
const amrCountData = find(["gene", geneLabel], sample.amr_counts)

// If the amr count entry doesn't exist, find() will return null. 
// Use get() to access the fields, since it can handle null values without throwing an error.
const depth = get("depth", amrCountData) || 0
const coverage = get("coverage", amrCountData) || 0
```

You can also use the `keyBy` function to key turn the array into an object, so that you no longer need to use the `find` function, which is O(n).

```
const sampleNameToAmrCounts = keyBy('sample_name', rawData)
```

I would recommend spending some time familiarizing yourself with `lodash/fp` and using it to simplify your logic here.

Reply:
Rewrote much of the logic to be more functional and removed the convoluted, unnecessary data structure. Thank you for your input!